src/Razor/src/Microsoft.VisualStudio.RazorExtension/SyntaxVisualizer/SyntaxVisualizerControl.xaml.cs

```diff
@@ -0,0 +1,467 @@
+            get
+            {
+                if (_runningDocumentTable == null)
+                {
+                    _runningDocumentTable = VSServiceHelpers.GetRequiredMefService<IVsRunningDocumentTable, SVsRunningDocumentTable>();
+                }
+
+                return _runningDocumentTable;
+            }
+        }
+
+        public SyntaxVisualizerControl()
+        {
+            InitializeComponent();
+
+            InitializeRunningDocumentTable();
+        }
+
+        [MemberNotNull(nameof(_codeDocumentProvider), nameof(_textDocumentFactoryService), nameof(_joinableTaskFactory))]
+        private void EnsureInitialized()
+        {
+            if (_codeDocumentProvider is not null && _textDocumentFactoryService is not null && _joinableTaskFactory is not null)
+            {
+                return;
+            }
+
+            _codeDocumentProvider = VSServiceHelpers.GetRequiredMefService<RazorCodeDocumentProvidingSnapshotChangeTrigger>();
+            _textDocumentFactoryService = VSServiceHelpers.GetRequiredMefService<ITextDocumentFactoryService>();
+            _joinableTaskFactory = VSServiceHelpers.GetRequiredMefService<JoinableTaskContext>().Factory;
+        }
+
+        private void InitializeRunningDocumentTable()
+        {
+            ThreadHelper.ThrowIfNotOnUIThread();
+
+            if (RunningDocumentTable != null)
+            {
+                RunningDocumentTable.AdviseRunningDocTableEvents(this, out _runningDocumentTableCookie);
+            }
+        }
+
+        void IDisposable.Dispose()
+        {
+            ThreadHelper.ThrowIfNotOnUIThread();
+
+            if (_runningDocumentTableCookie != 0)
+            {
+                _runningDocumentTable?.UnadviseRunningDocTableEvents(_runningDocumentTableCookie);
+                _runningDocumentTableCookie = 0;
+            }
+        }
+
+        private void ShowGeneratedCode_Click(object sender, RoutedEventArgs e)
+        {
+            if (_activeWpfTextView is null)
+            {
+                return;
+            }
+
+            EnsureInitialized();
+
+            var textBuffer = _activeWpfTextView.TextBuffer;
+
+            if (!_textDocumentFactoryService.TryGetTextDocument(textBuffer, out var textDocument))
+            {
+                return;
+            }
+
+            var codeDocument = _joinableTaskFactory.Run(() => _codeDocumentProvider.GetRazorCodeDocumentAsync(textDocument.FilePath, CancellationToken.None));
+            if (codeDocument is null)
+            {
+                return;
+            }
+
+            var fileName = Path.GetFileName(textDocument.FilePath);
+            var tempFileName = Path.Combine(Path.GetTempPath(), fileName + ".g.cs");
+
+            // Ignore any I/O errors
+            try
+            {
+                File.WriteAllText(tempFileName, codeDocument.GetCSharpDocument().GeneratedCode);
+                VsShellUtilities.OpenDocument(ServiceProvider.GlobalProvider, tempFileName);
```

Comment:
This seems fishy to me. Is there really no way to open a "virtual" document in VS? Seems more performant and less error prone than writing it to disk (and then never cleaning it up).

Reply:
There really is no way. In GTD for embedded source I have to write temp files. If you open the source generated code for Razor from the Analyzers node in solution explorer, it will create a temp file.